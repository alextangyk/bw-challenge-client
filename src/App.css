.App {
  /* text-align: center; */
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

.Topic {
  color: gray;
}

.Topic-Green {
  color: green;
}

.Topic-Red {
  color: red;
}

.Topic-Font1 {
  font-size: 16px;
}

.Topic-Font2 {
  font-size: 20px;
}

.Topic-Font3 {
  font-size: 24px;
}

.Topic-Font4 {
  font-size: 28px;
}

.Topic-Font5 {
  font-size: 32px;
}

.Topic-Font6 {
  font-size: 36px;
}

.Container {
  display: flex;
  flex-direction: row;
}

.Col {
  flex: 1;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
