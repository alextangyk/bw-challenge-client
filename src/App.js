import React, { useEffect, useState } from 'react';
import axios from 'axios';

import './App.css';

function App() {
  const [topics, setTopics] = useState([]);
  const [selectedTopic, setSelectedTopic] = useState('');

  useEffect(() => {
    const getTopicsApi = async () => {
      try {
        const response = await axios.get(`${process.env.REACT_APP_API_URL}/api/topics`);
        const topics = response.data;
        setTopics(topics);
      } catch (error) { }
    }
    getTopicsApi()
  }, []);

  const selectTopicMeta = topics.find(t => t.label === selectedTopic);

  const positive = selectTopicMeta?.sentiment?.positive || 0;
  const neutral = selectTopicMeta?.sentiment?.neutral || 0;
  const negative = selectTopicMeta?.sentiment?.negative || 0;
  const total = selectTopicMeta?.volume;

  const volumes = topics.map(t => t.volume);
  const minVolume = volumes.length > 0 ? Math.min(...volumes) : -1;
  const maxVolume = volumes.length > 0 ? Math.max(...volumes) : -1;
  const division = (maxVolume - minVolume) / 6;

  return (
    <div className="App">
      <h1>
        My Topics Challenge
      </h1>
      <div className="Container">
        <div className="Col">
          {
            topics.map(topic => {
              let colorClassName = "Topic";
              if (topic.sentimentScore > 60) {
                colorClassName = "Topic-Green";
              } else if (topic.sentimentScore < 40) {
                colorClassName = "Topic-Red";
              }

              let size = 1;
              if (minVolume > 0 && maxVolume > 0) {
                const checkValue = topic.volume - minVolume;

                for (let i = 0; i < 6; i += 1) {
                  if (checkValue === 0 || 
                    (checkValue > i * division && checkValue <= ((i + 1) * division))) {
                    size = i + 1;
                    break;
                  }
                }
              }

              return (
                <div 
                  key={topic.id} 
                  data-testid={topic.id} 
                  className={`${colorClassName} Topic-Font${size}`} 
                  onClick={() => setSelectedTopic(topic.label)}
                >
                  {topic.label}
                </div>
              );
            })
          }
        </div>
        {selectTopicMeta && <div className="Col">
          <div>Infomation of Topic <span>{`"${selectedTopic}"`}</span></div>
          <br />
          <div>Total Mentions:<span data-testid="total">{` ${total}`}</span></div>
          <br />
          <div>Positive Mentions:<span data-testid="positive" className="Topic-Green">{` ${positive}`}</span></div>
          <div>Neutral Mentions:<span data-testid="neutral">{` ${neutral}`}</span></div>
          <div>Negative Mentions:<span data-testid="negative" className="Topic-Red">{` ${negative}`}</span></div>
        </div>}

      </div>

    </div>
  );
}

export default App;
