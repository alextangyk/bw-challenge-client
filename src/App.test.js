import { render, screen, fireEvent } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import axios from 'axios';
import App from './App';

describe('Testing App', () => {
  const responseData = [
    {
      "id": "1751295897__Berlin",
      "label": "Berlin",
      "volume": 165,
      "type": "topic",
      "sentiment": {
        "negative": 3,
        "neutral": 133,
        "positive": 29
      },
      "sentimentScore": 65,
      "burst": 13,
      "days": [
        {
          "date": "2014-06-06T00:00:00.000+0000",
          "volume": 22
        },
        {
          "date": "2014-06-04T00:00:00.000+0000",
          "volume": 43
        },
        {
          "date": "2014-06-09T00:00:00.000+0000",
          "volume": 0
        },
        {
          "date": "2014-06-07T00:00:00.000+0000",
          "volume": 12
        },
        {
          "date": "2014-06-08T00:00:00.000+0000",
          "volume": 11
        },
        {
          "date": "2014-06-03T00:00:00.000+0000",
          "volume": 39
        },
        {
          "date": "2014-06-05T00:00:00.000+0000",
          "volume": 38
        }
      ],
      "pageType": {
        "blog": 17,
        "facebook": 56,
        "forum": 22,
        "general": 5,
        "image": 0,
        "news": 26,
        "review": 1,
        "twitter": 35,
        "video": 3
      },
      "queries": [
        {
          "id": 1751295897,
          "name": "Berghain",
          "volume": 165
        }
      ]
    },
    {
      "id": "1751295897__Quantified Drunk",
      "label": "Quantified Drunk",
      "volume": 14,
      "type": "topic",
      "sentiment": {
        "neutral": 14
      },
      "sentimentScore": 50,
      "burst": 7,
      "days": [
        {
          "date": "2014-06-06T00:00:00.000+0000",
          "volume": 1
        },
        {
          "date": "2014-06-04T00:00:00.000+0000",
          "volume": 8
        },
        {
          "date": "2014-06-09T00:00:00.000+0000",
          "volume": 0
        },
        {
          "date": "2014-06-07T00:00:00.000+0000",
          "volume": 0
        },
        {
          "date": "2014-06-08T00:00:00.000+0000",
          "volume": 1
        },
        {
          "date": "2014-06-03T00:00:00.000+0000",
          "volume": 0
        },
        {
          "date": "2014-06-05T00:00:00.000+0000",
          "volume": 4
        }
      ],
      "pageType": {
        "blog": 0,
        "facebook": 3,
        "forum": 0,
        "general": 0,
        "image": 0,
        "news": 0,
        "review": 0,
        "twitter": 11,
        "video": 0
      },
      "queries": [
        {
          "id": 1751295897,
          "name": "Berghain",
          "volume": 14
        }
      ]
    },
    {
      "id": "1751295897__Hammered",
      "label": "Hammered",
      "volume": 48,
      "type": "topic",
      "sentiment": {
        "neutral": 18,
        "negative": 30
      },
      "sentimentScore": 20,
      "burst": 5,
      "days": [
        {
          "date": "2014-06-06T00:00:00.000+0000",
          "volume": 1
        },
        {
          "date": "2014-06-04T00:00:00.000+0000",
          "volume": 8
        },
        {
          "date": "2014-06-09T00:00:00.000+0000",
          "volume": 0
        },
        {
          "date": "2014-06-07T00:00:00.000+0000",
          "volume": 0
        },
        {
          "date": "2014-06-08T00:00:00.000+0000",
          "volume": 1
        },
        {
          "date": "2014-06-03T00:00:00.000+0000",
          "volume": 0
        },
        {
          "date": "2014-06-05T00:00:00.000+0000",
          "volume": 8
        }
      ],
      "pageType": {
        "blog": 0,
        "facebook": 3,
        "forum": 0,
        "general": 0,
        "image": 0,
        "news": 0,
        "review": 0,
        "twitter": 15,
        "video": 0
      },
      "queries": [
        {
          "id": 1751295897,
          "name": "Berghain",
          "volume": 18
        }
      ]
    },
  ];

  let axiosGetSpy = null;

  beforeEach(() => {
    axiosGetSpy = jest.spyOn(axios, 'get').mockResolvedValueOnce({ data: responseData });
  });

  afterEach(() => {
    axiosGetSpy.mockClear();
  })

  test('render correctly', () => {
    const tree = TestRenderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('get topic color', async () => {
    const { findByTestId } = render(<App />);

    expect(axiosGetSpy).toBeCalledWith(`${process.env.REACT_APP_API_URL}/api/topics`);

    const topic1 = await findByTestId(responseData[0].id);
    expect(topic1).toBeTruthy();
    expect(topic1).toHaveClass('Topic-Green');
    expect(topic1).toHaveClass('Topic-Font6');

    const topic2 = await findByTestId(responseData[1].id);
    expect(topic2).toBeTruthy();
    expect(topic2).toHaveClass('Topic');
    expect(topic2).toHaveClass('Topic-Font1');

    const topic3 = await findByTestId(responseData[2].id);
    expect(topic3).toBeTruthy();
    expect(topic3).toHaveClass('Topic-Red');
    expect(topic3).toHaveClass('Topic-Font2');
  });

  test('click topic', async () => {
    const { findByTestId, findByText } = render(<App />);

    expect(axiosGetSpy).toBeCalledWith(`${process.env.REACT_APP_API_URL}/api/topics`);

    const topic = await findByTestId(responseData[0].id);
    expect(topic).toBeTruthy();

    fireEvent.click(topic);

    const negative = await findByText(`Negative Mentions:`);
    const negativeContent = await findByTestId('negative');
    const negativeValue = responseData[0].sentiment.negative;
    expect(negative).toBeTruthy();
    expect(negativeContent.textContent).toBe(` ${negativeValue}`);

    const positive = await findByText('Positive Mentions:');
    const positiveContent = await findByTestId('positive');
    const positiveValue = responseData[0].sentiment.positive;
    expect(positive).toBeTruthy();
    expect(positiveContent.textContent).toBe(` ${positiveValue}`);

    const neutral = await findByText('Neutral Mentions:');
    const neutralContent = await findByTestId('neutral');
    const neutralValue = responseData[0].sentiment.neutral;
    expect(neutral).toBeTruthy();
    expect(neutralContent.textContent).toBe(` ${neutralValue}`);

    const totalContent = await findByTestId('total');
    const totalValue = responseData[0].volume;
    expect(totalContent.textContent).toBe(` ${totalValue}`);
  })
})
