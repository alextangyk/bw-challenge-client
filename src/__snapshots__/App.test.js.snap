// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing App render correctly 1`] = `
<div
  className="App"
>
  <h1>
    My Topics Challenge
  </h1>
  <div
    className="Container"
  >
    <div
      className="Col"
    />
  </div>
</div>
`;
